crd-generator/src/main/java/io/fabric8/crd/generator/decorator/NamedResourceDecorator.java

```diff
@@ -0,0 +1,129 @@
+/**
+ * Copyright (C) 2015 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.fabric8.crd.generator.decorator;
+
+import static io.fabric8.crd.generator.utils.Metadata.getKind;
+import static io.fabric8.crd.generator.utils.Metadata.getMetadata;
+
+import io.fabric8.crd.generator.utils.Generics;
+import io.fabric8.crd.generator.utils.Strings;
+import io.fabric8.kubernetes.api.builder.TypedVisitor;
+import io.fabric8.kubernetes.api.builder.VisitableBuilder;
+import io.fabric8.kubernetes.api.model.ObjectMeta;
+import java.util.Optional;
+
+public abstract class NamedResourceDecorator<T> extends Decorator<VisitableBuilder> {
+
+  /**
+   * For resource name null acts as a wildcards.
+   * Let's use a constant instead, for clarity's shake
+   */
+  public static final String ANY = null;
+
+  protected final String kind;
+  protected final String name;
+
+  private final ResourceVisitor visitor = new ResourceVisitor(null, null);
+
+  public NamedResourceDecorator() {
+    this(ANY, ANY);
+  }
+
+  public NamedResourceDecorator(String name) {
+    this(ANY, name);
+  }
+
+  public NamedResourceDecorator(String kind, String name) {
+    this.kind = kind;
+    this.name = name;
+  }
+
+  @Override
+  public void visit(VisitableBuilder builder) {
+    Optional<String> resourceKind = getKind(builder);
+    Optional<ObjectMeta> objectMeta = getMetadata(builder);
+    if (!resourceKind.isPresent() || !objectMeta.isPresent()) {
+      return;
+    }
+    if (Strings.isNullOrEmpty(kind)) {
+      if (Strings.isNullOrEmpty(name)) {
+        builder.accept(visitor.withKind(resourceKind.get()).withMetadata(objectMeta.get()));
+      } else if (objectMeta.map(m -> m.getName()).filter(s -> s.equals(name)).isPresent()) {
+        builder.accept(visitor.withKind(resourceKind.get()).withMetadata(objectMeta.get()));
+      }
+    } else if (resourceKind.filter(k -> k.equals(kind)).isPresent()) {
+      if (Strings.isNullOrEmpty(name)) {
+        builder.accept(visitor.withKind(resourceKind.get()).withMetadata(objectMeta.get()));
+      } else if (objectMeta.map(m -> m.getName()).filter(s -> s.equals(name)).isPresent()) {
+        builder.accept(visitor.withKind(resourceKind.get()).withMetadata(objectMeta.get()));
+      }
```

Comment:
Umm, why are we executing the same code in both blocks?

Please check code smells reported by sonar and address whatever seem relevant to you.

Reply:
Good question! As far as I'm aware I ported that code as-is from Dekorate. Any idea why, @iocanel?

Reply:
No particular reason, most probably it never occurred to me that that they were identical.

Since the logic here is: `if kind is null or matching do <stuff>` I guess that it would make sense to collapse both parts into a single if.